Adds demo setup and instructions

lukpueh opened the pull request:
Adds a demo directory with project owner and functionary keys and instructions on how to download and install in-toto and go through a simple software supply chain. This should replace the previous demo instructions that required the user to download a linux VM.

lukpueh committed:
Add demo folders and key pairs for demo roles

lukpueh committed:
Updates demo

- Adds README with detailed demo instructions
- Adds create_layout.py to create simple layout
- Updates requirements.txt
- Renames demo to run_demo and adds it into demo folder

lukpueh committed:
Adds dummy file to empty demo folder

Git does not keep track of empfy dirs but we want this dir in
place for demo purposes

lukpueh committed:
Fixes demo instructions

- Fixes some minor demo README nits

lukpueh committed:
Updates run_demo.py to work with new instructions

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+workflow.
+Alice will be the project owner - she creates and signs the supply chain layout
+with her private key - and Bob and Carl will be the project functionaries - they
+carry out the steps of the supply chain as defined in the layout.
```
Can you include a few more sentences to explain what is meant by "supply chain?" (note: consider using "software supply chain")  Likewise, it would help the introduction if "layout" were defined.  What does the layout for this simple workflow resemble?

You will likely cover in-toto's design in a presentation, but it would be beneficial to the reader to have a README of sorts for this.  The current [README](https://github.com/in-toto/in-toto/blob/master/README.md) doesn't include an overview of in-toto.  Perhaps you could link the specification here?

lukpueh replied:
Sure. I guess the repo's README is a better place for high level descriptions and terminology.
@SantiagoTorres do you want to give it a shot?

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+
+# Install additional requirements that for some good reason are not in the
+# requirements file
+pip install pycrypto, cryptography
```
You'd get an error if this command were executed.  I think the comma shouldn't be there.

SantiagoTorres commented:
On Mon, Oct 31, 2016 at 09:04:20AM -0700, lukpueh wrote:

> lukpueh commented on this pull request.
> 
> > @@ -0,0 +1,203 @@
> > +# in-toto demo
> > +
> > +In this demo, we will use in-toto to secure a supply chain with a very simple
> > +workflow.
> > +Alice will be the project owner - she creates and signs the supply chain layout
> > +with her private key - and Bob and Carl will be the project functionaries - they
> > +carry out the steps of the supply chain as defined in the layout.
> 
> Sure. I guess the repo's README is a better place for high level descriptions and terminology.
> @SantiagoTorres do you want to give it a shot?

Sure, I can give this a shot. I think we primarly want to move things
around from the website into the README for this iteration.

Thanks,
-Santiago.

> ## 
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/in-toto/in-toto/pull/28

lukpueh committed:
Removes comma from pip install in demo README

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+
+# Make sure in-toto is on your PYTHONPATH and used commands are on your PATH
+export PYTHONPATH=$PYTHONPATH:$PWD
+export PATH=$PATH:$PWD/toto
```
Is this only needed so that the reader only has to enter the script name on the command line?  Consider using the [scripts keyword in setup.py](https://docs.python.org/2/distutils/setupscript.html#installing-scripts).  You might need to include a shebang on the first line of these scripts.

Also, if students will be contributing code it can useful to install in-toto in "editable" mode (add `--editable .` to the requirements.txt file).

lukpueh replied:
Yep, but I wasn't sure if we had the setup.py merged until tomorrow.

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+# ├── owner_alice
+# │   ├── alice
+# │   └── alice.pub
+# └── run-demo.py
```
This is what I see (note: create_layout.py is missing from the listed output)

``` Bash
$ tree
.
├── final_product
├── functionary_bob
│   ├── bob
│   └── bob.pub
├── functionary_carl
│   ├── carl
│   └── carl.pub
├── owner_alice
│   ├── alice
│   ├── alice.pub
│   └── create_layout.py
├── README.md
└── run_demo.py

4 directories, 9 files
```

SantiagoTorres commented:
On Mon, Oct 31, 2016 at 09:49:41AM -0700, lukpueh wrote:

> lukpueh commented on this pull request.
> 
> > +git clone -b develop --recursive https://github.com/in-toto/in-toto.git
> > +
> > +# Change into project root directory
> > +cd in-toto
> > +
> > +# Install required packages (we strongly recommend using Virtual Environments)
> > +# http://docs.python-guide.org/en/latest/dev/virtualenvs/
> > +pip install -r requirements.txt
> > +
> > +# Install additional requirements that for some good reason are not in the
> > +# requirements file
> > +pip install pycrypto, cryptography
> > +
> > +# Make sure in-toto is on your PYTHONPATH and used commands are on your PATH
> > +export PYTHONPATH=$PYTHONPATH:$PWD
> > +export PATH=$PATH:$PWD/toto
> 
> Yep, but I wasn't sure if we had the setup.py merged until tomorrow.

Why? This PR has been up there for a week now and it's rather trivial.

> ## 
> 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/in-toto/in-toto/pull/28

vladimir-v-diaz commented on demo/README.md:
````diff
@@ -0,0 +1,203 @@
+```shell
+# Create and sign the supply chain layout on behalf of Alice
+cd owner_alice
+python create_layout.py
````
I get the following exception when running create_layout.py:

``` Bash
$ python create_layout.py 
Traceback (most recent call last):
  File "create_layout.py", line 1, in <module>
    from toto.util import import_rsa_key_from_file
  File "/tmp/in-toto/toto/util.py", line 7, in <module>
    import toto.ssl_crypto.keys
  File "/tmp/in-toto/toto/ssl_crypto/keys.py", line 93, in <module>
    from . import pycrypto_keys as ssl_crypto__pycrypto_keys
  File "/tmp/in-toto/toto/ssl_crypto/pycrypto_keys.py", line 153, in <module>
    _PBKDF2_ITERATIONS = settings.PBKDF2_ITERATIONS
  File "/tmp/intoto-env/local/lib/python2.7/site-packages/simple_settings/core.py", line 84, in __getattr__
    self.setup()
  File "/tmp/intoto-env/local/lib/python2.7/site-packages/simple_settings/core.py", line 62, in setup
    raise RuntimeError('Settings are not configured')
RuntimeError: Settings are not configured
```

lukpueh replied:
Good catch! Forgot to add `export SIMPLE_SETTINGS=toto.settings`

vladimir-v-diaz replied:
I recommend that you set the general cryptography library (settings.GENERAL_CRYPTO_LIBRARY) to `pyca-cryptography`, which is faster and more solid than `pycrypto`.  We recently (past 2 weeks) set it to `pyca-cryptography` in TUF's settings, so that might be why it's not set here.

vladimir-v-diaz commented on demo/README.md:
````diff
@@ -0,0 +1,203 @@
+
+```shell
+cd ../functionary_bob
+toto-run.py --step-name write-code --products foo.py --key bob -- vi foo.py
````
I get an exception for this line.

``` Bash
$ toto-run.py --step-name write-code --products foo.py --key bob -- vi foo.py
Traceback (most recent call last):
  File "/tmp/in-toto/toto/toto-run.py", line 36, in <module>
    import toto.runlib
  File "/tmp/in-toto/toto/runlib.py", line 34, in <module>
    import subprocess32 as subprocess
ImportError: No module named subprocess32
```

vladimir-v-diaz replied:
There is an `optional-requirements.txt` file that includes subprocess32.  Should this exception explain that POSIX users are encouraged to `pip install subprocess32` ?

vladimir-v-diaz replied:
Here's my output after installing `subprocess32`and successfully running `toto-run.py`:

``` Bash
$ toto-run.py --step-name write-code --products foo.py --key bob -- vi foo.py
DOING:   load link signing key...
DOING:   record materials...
DOING:   run command...
DOING:   record products...
DOING:   create link metadata...
DOING:   sign metadata...
DOING:   store metadata...
```

lukpueh replied:
I have added a fallback in https://github.com/in-toto/in-toto/pull/27. This is already merged into develop. I guess we need another round of rebase before we merge add-demo.

vladimir-v-diaz commented on demo/README.md:
````diff
@@ -0,0 +1,203 @@
+cp foo.py ../functionary_carl/
+```
+The command you just entered will open a `vi` editor, where you can write your
+code (you can write whatever you want). After you save the file and close vi
````
I guess you'll cover this later, but what if the code has already been written?  How would this command change?  (maybe something to mention for the curious reader.)

lukpueh replied:
Read the specs! :D

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+cd ../functionary_bob
+toto-run.py --step-name write-code --products foo.py --key bob -- vi foo.py
+# Bob has to send the resulting foo.py to Carl so that he can package it
+cp foo.py ../functionary_carl/
```
I'm not sure if others are likely to follow my behaviour... but after executing toto-run.py and inspecting the output, I quickly jumped to the text explaining the command/output.  I missed/overlooked the `cp foo.py ...`  It might make more sense to stick this command in the next section, where Package Happy Carl performs his step.

vladimir-v-diaz replied:
Or put a blank line between the two commands:

``` Python
cd ../functionary_bob
toto-run.py --step-name write-code --products foo.py --key bob -- vi foo.py

# Bob has to send the resulting foo.py to Carl so that he can package it
cp foo.py ../functionary_carl/
```

vladimir-v-diaz commented on demo/README.md:
````diff
@@ -0,0 +1,203 @@
+
+```shell
+cd ../functionary_carl
+toto-run.py --step-name package --materials foo.py --products foo.tar.gz --key carl -- tar zcvf foo.tar.gz foo.py
````
Here is my output.  The stray `foo.py` looks a bit odd.

``` Bash
DOING:   load link signing key...
DOING:   record materials...
DOING:   run command...
foo.py
DOING:   record products...
DOING:   create link metadata...
DOING:   sign metadata...
DOING:   store metadata...
```

lukpueh commented:
I agree, it does.  But it's the actual stdout from the `tar` command. Any ideas?

vladimir-v-diaz commented on demo/README.md:
````diff
@@ -0,0 +1,203 @@
+And now run verification on behalf of the client:
+```shell
+cd final_product
+# Fetch Alices public key from a trusted source to verify the layout signature
````
Minor typo: Alice's

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+# Fetch Alices public key from a trusted source to verify the layout signature
+# Note: The functionary public keys are fetched from the layout
+cp ../owner_alice/alice.pub .
+toto-verify.py --layout root.layout --layout-key alice.pub
```
My output: 

``` Bash
DOING:   load layout...
DOING:   verify layout expiration
DOING:   load layout keys...
DOING:   verify layout signatures...
DOING:   load link metadata for steps...
DOING:   verify signatures for all links...
DOING:   run all inspections...
DOING:   verify all step matchrules...
DOING:   verify all inspection matchrules...
PASSING: all verification
```

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+ 1. that each step was performed and signed by the authorized functionary,
+ 1. that the functionaries used the commands they were supposed to use (`vi`, `tar`) and
+ 1. that the recorded materials and product align with the matchrules
+ 1. that the defined inspection `untar` finds what it expects.
```
The reader did not design `root.layout` and define `untar` themselves, so might be worthwhile to mention here that it's defined in root.layout.

vladimir-v-diaz commented on demo/README.md:
````diff
@@ -0,0 +1,203 @@
+
+```shell
+cd final_product
+toto-verify.py --layout root.layout --layout-key alice.pub
````
My output: 

``` Bash
DOING:   load layout...
DOING:   verify layout expiration
DOING:   load layout keys...
DOING:   verify layout signatures...
DOING:   load link metadata for steps...
DOING:   verify signatures for all links...
DOING:   run all inspections...
DOING:   verify all step matchrules...
FAILING: in verify all step matchrules - hashes of 'foo.py' do not match 
```

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+
+
+### Wrapping up
+Congratulations! You have completed the Toto demo! This exercise shows a very
```
Toto -> in-toto

SantiagoTorres replied:
bump, we still need to change this right?

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+Congratulations! You have completed the Toto demo! This exercise shows a very
+simple case in how Toto can protect the different steps within the supply chain.
+More complex supply chains that contain more steps can be created in a similar
+way. You can read more about what Toto protects against and how to use it here.
```
Is the `here` supposed to be a link?

SantiagoTorres replied:
Yes, this should be a link.

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+### Wrapping up
+Congratulations! You have completed the Toto demo! This exercise shows a very
+simple case in how Toto can protect the different steps within the supply chain.
+More complex supply chains that contain more steps can be created in a similar
```
Recommend that all instances of `supply chain` include "software"

SantiagoTorres replied:
bump here too

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+### Tired of copy-pasting commands?
+We provide a `run_demo.py` script that sequentially executes all commands
+listed above. Just change into the `demo` directory, run it and observe the
+output.
```
Now you tell me! :dagger:

vladimir-v-diaz commented on demo/README.md:
```diff
@@ -0,0 +1,203 @@
+# in-toto demo
```
A table of contents might be cool to have.

lukpueh committed:
Addresses Vlad's and Santiago's comments

- Replaces pip -r with pip -e (using setup.py)
- Removes setting PATH and PYTHONPATH
- Adds missing SIMPLE_SETTINGS envvar export
- Adds missing file in tree output
- Separates "send file to carl" step
- Clarifies behind the scenes of toto-verify

SantiagoTorres commented on demo/owner_alice/create_layout.py:
```diff
@@ -0,0 +1,54 @@
+  layout.dump()
+
+if __name__ == '__main__':
+  main()
```
missing newline

SantiagoTorres commented on demo/README.md:
```diff
@@ -0,0 +1,210 @@
+This command will verify that
+ 1. the layout has not expired,
+ 2. was signed with Alice’s private key,
+<br>and that according to the definitions in the layout
```
this newline fits? does it render?

lukpueh replied:
It's a little layout hack. I guess it can be done nicer.

SantiagoTorres commented on toto/toto-verify.py:
```diff
@@ -47,6 +46,12 @@ def in_toto_verify(layout_path, layout_key_paths):
+  except Exception, e:
+    _die("in verify layout expiration - %s" % e)
+
+  try:
```
Hmm, what's with these changes? are they related to this PR?

SantiagoTorres commented on toto/verifylib.py:
```diff
@@ -70,6 +73,30 @@ def run_all_inspections(layout):
+    # raise LayoutExpiredError
+    raise Exception("Layout expired")
+
+
```
Shouldn't this have been on #24?

SantiagoTorres reviewed (changes requested):
Minor nits, I just want to make sure everything is in place before merging.

lukpueh committed:
Addresses more of Vlad's comments

- Replaces toto with in-toto
- Replaces supply chain with software supply chain
- Adds missing link

SantiagoTorres commented:
Thanks! :100: :beer:

lukpueh commented:
Kudos to the reviewer @SantiagoTorres and @vladimir-v-diaz!!
